File: Greedy/PartitionLabels.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> partitionLabels(string s) {
        vector<int> ans;
        vector<vector<int>> intervals;
        unordered_map<char, vector<int>> indexes;
        for (int i = 0; i < s.size(); i++) {
            if (indexes.count(s[i]) == 0) {
                indexes[s[i]] = vector<int>();
            }
            indexes[s[i]].push_back(i);
        }
        for (auto& [key, vec] : indexes) {
            intervals.push_back({vec.front(), vec.back()});
        }
        sort(intervals.begin(), intervals.end());
        int minI = intervals[0][0];
        int maxI = intervals[0][1];
        int i = 1;
        while (i < intervals.size()) {
            if (intervals[i][0] > maxI) {
                ans.push_back(maxI - minI + 1);
                minI = intervals[i][0];
                maxI = intervals[i][1];
            }
            else if (intervals[i][0] < maxI) {
                if (intervals[i][1] > maxI) {
                    maxI = intervals[i][1];
                }
            }
            if (i == intervals.size() - 1) {
                ans.push_back(maxI - minI + 1);
            }
            i += 1;
        }
        return ans;
    }
};

File: Greedy/PartitionLabels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        ans = []
        intervals = []
        indexes = {}
        for i in range(len(s)):
            if s[i] not in indexes:
                indexes[s[i]] = []
            indexes[s[i]].append(i)
        for k in indexes:
            intervals.append([min(indexes[k]), max(indexes[k])])
        intervals.sort(key=self.getKey)
        minI = intervals[0][0]
        maxI = intervals[0][1]
        i = 1
        while i < len(intervals):
            if intervals[i][0] > maxI:
                ans.append(maxI - minI + 1)
                minI = intervals[i][0]
                maxI = intervals[i][1]
            elif intervals[i][0] < maxI:
                if intervals[i][1] > maxI:
                    maxI = intervals[i][1]
            if i == len(intervals) - 1:
                ans.append(maxI - minI + 1)
                minI = intervals[i][0]
                maxI = intervals[i][1]
            i += 1
        return ans

    def getKey(self, key):
        return key[0]


